operator/v1/types_machineconfiguration.go

```diff
@@ -153,3 +167,181 @@ const (
+	// userPolicies define user-provided node disruption policies
+	// +optional
+	UserPolicies NodeDisruptionPolicyConfig `json:"userPolicies"`
+
+	// clusterDefaultPolicies is managed by the Machine Config Operator, and reflects the latest cluster defaults
+	// +optional
+	ClusterDefaultPolicies NodeDisruptionPolicyConfig `json:"clusterDefaultPolicies"`
+}
+
+type NodeDisruptionPolicyStatus struct {
+	// clusterPolicies is a merge of cluster default and user provided node disruption policies.
+	// +optional
+	ClusterPolicies NodeDisruptionPolicyClusterStatus `json:"clusterPolicies"`
+
+	// conditions represent the current validation status of the NodeDisruptionPolicy in spec
+	// +patchMergeKey=type
+	// +patchStrategy=merge
+	// +listType=map
+	// +listMapKey=type
+	Conditions []metav1.Condition `json:"conditions,omitempty" patchStrategy:"merge" patchMergeKey:"type"`
+}
+
+// NodeDisruptionPolicyConfig is the overall spec definition for files/units/sshkeys
+type NodeDisruptionPolicyConfig struct {
+	// files is a list of MachineConfig file definitions and actions to take to changes on those paths
+	// +optional
+	// +listType=map
+	// +listMapKey=path
+	Files []NodeDisruptionPolicyFile `json:"files,omitempty"`
+	// units is a list MachineConfig unit definitions and actions to take on changes to those services
+	// +optional
+	// +listType=map
+	// +listMapKey=name
+	Units []NodeDisruptionPolicyUnit `json:"units,omitempty"`
+	// sshkey is the overall sshkey MachineConfig definition
+	// +optional
+	SSHKey NodeDisruptionPolicySSHKey `json:"sshkey,omitempty"`
+}
+
+// NodeDisruptionPolicyClusterStatus is the type for the status object, rendered by the controller as a
+// merge of cluster defaults and user provided policies
+type NodeDisruptionPolicyClusterStatus struct {
+	// files is a list of MachineConfig file definitions and actions to take to changes on those paths
+	// +optional
+	// +listType=map
+	// +listMapKey=path
+	Files []NodeDisruptionPolicyFile `json:"files,omitempty"`
+	// units is a list MachineConfig unit definitions and actions to take on changes to those services
+	// +optional
+	// +listType=map
+	// +listMapKey=name
+	Units []NodeDisruptionPolicyUnit `json:"units,omitempty"`
+	// sshkey is the overall sshkey MachineConfig definition
+	// +optional
+	SSHKey NodeDisruptionPolicySSHKey `json:"sshkey,omitempty"`
+}
+
+// NodeDisruptionPolicyFile is a file entry and corresponding actions to take
+type NodeDisruptionPolicyFile struct {
+	// path represents the filepath
+	// +kubebuilder:validation:Required
+	Path string `json:"path"`
+	// actions represents the series of commands to be executed on changes to the corresponding type and value
+	// +kubebuilder:validation:Required
+	// +listType=atomic
+	Actions []NodeDisruptionPolicyAction `json:"actions"`
+}
+
+// NodeDisruptionPolicyUnit is a systemd unit name and corresponding actions to take
+type NodeDisruptionPolicyUnit struct {
+	// name represents the service name
```

Comment:
Ignition does apply a valid name check it looks like: https://github.com/coreos/ignition/blob/d84b20fadef7cedfd3577f27272c89b5ddb0f4d2/config/v3_4/types/unit.go#L48-L55

Should we do that here? I was thinking of having the MCO validation take care of that

Reply:
Ideally, we apply as much validation as possible at admission time, so that we never persist an invalid object. If it's possible to validate this name here (which it looks like it is), then we should aim to do so.

We prefer openapi validations (eg CEL and kubebuilder patterns), but validating webhooks can also be used for more complex validations. But the ideal world is that all validations happen at admission, and are not "Check by controller, set status bad", the latter pattern meaning a user can apply the object, get a success and walk away, thinking they've achieved something that really they have not.

Reply:
Ack, will try to construct a validator, thanks!